File: MySolutions/Python/q8_Atoi.py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        i=0
        n=len(s)
        sign=1
        res=0

        while i<n and s[i]==" ":  #skipping whitespaces
            i+=1

        if i<n and (s[i]=='-' or s[i]=='+'):  #defining sign
            if s[i]=='-':
                sign = -1
            i+=1

        while i<n and '0'<=s[i]<='9':  #actual conversion
            res = res*10 + (ord(s[i])-ord('0'))
            i+=1
        
        res*=sign  #ensuring sign is included
        
        maxlim = 2**31 - 1
        minlim = -2**31

      #checking for out of bounds cases
        if res > maxlim:  
            return maxlim
        if res < minlim:
            return minlim
        
        return res        

File: MySolutions/Cpp/q8_Atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int sign=1;
        long long int res=0;
        int n=s.size();
        int i=0;
        
        while(i<n && s[i]==' '){
            i++;
        }
        
        if(i<n && (s[i]=='-'||s[i]=='+')){
            if(s[i]=='-'){
                sign=-1;
            }
            i++;
        }
        
        while(i<n && '0'<=s[i] && s[i]<='9'){

             if (res > INT_MAX / 10 || (res == INT_MAX / 10 && s[i] - '0' > INT_MAX % 10)) {
                return sign == 1 ? INT_MAX : INT_MIN;
            }
            
            res = res*10 + (s[i] - '0');
            i++;
        }
        res*=sign;
        
        if(res>INT_MAX) res=INT_MAX;
        else if(res<INT_MIN) res=INT_MIN;
        
        return res;
    }
};
